core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java

```diff
@@ -360,4 +396,170 @@ private TableMetadata getTableMetadata(Operation operation) throws CrudException
   public Snapshot getSnapshot() {
     return snapshot;
   }
+
+  private interface ConsensusCommitScanner extends TransactionCrudOperable.Scanner {
+    boolean isClosed();
+  }
+
+  @NotThreadSafe
+  private class ConsensusCommitStorageScanner extends AbstractTransactionCrudOperableScanner
+      implements ConsensusCommitScanner {
+
+    private final Scan scan;
+    private final List<String> originalProjections;
+    private final Scanner scanner;
+
+    private final LinkedHashMap<Snapshot.Key, TransactionResult> results = new LinkedHashMap<>();
+    private boolean fullyScanned;
+    private boolean failed;
+    private boolean closed;
+
+    public ConsensusCommitStorageScanner(Scan scan, List<String> originalProjections)
+        throws CrudException {
+      this.scan = scan;
+      this.originalProjections = originalProjections;
+      scanner = scanFromStorage(scan);
+    }
+
+    @Override
+    public Optional<Result> one() throws CrudException {
+      try {
+        Optional<Result> r = scanner.one();
+
+        if (!r.isPresent()) {
+          fullyScanned = true;
+          return Optional.empty();
+        }
+
+        Snapshot.Key key = new Snapshot.Key(scan, r.get());
+        TransactionResult result = new TransactionResult(r.get());
+        processScanResult(key, scan, result);
+        results.put(key, result);
+
+        TableMetadata metadata = getTableMetadata(scan);
+        return Optional.of(
+            new FilteredResult(result, originalProjections, metadata, isIncludeMetadataEnabled));
+      } catch (ExecutionException e) {
+        failed = true;
+        throw new CrudException(
+            CoreError.CONSENSUS_COMMIT_SCANNING_RECORDS_FROM_STORAGE_FAILED.buildMessage(),
+            e,
+            snapshot.getId());
+      } catch (CrudException e) {
+        failed = true;
+        throw e;
+      }
+    }
+
+    @Override
+    public List<Result> all() throws CrudException {
+      List<Result> results = new ArrayList<>();
+
+      while (true) {
+        Optional<Result> result = one();
+        if (!result.isPresent()) {
+          break;
+        }
+        results.add(result.get());
+      }
+
+      return results;
+    }
+
+    @Override
+    public void close() {
+      if (closed) {
+        return;
+      }
+      closed = true;
+
+      try {
+        scanner.close();
+      } catch (IOException e) {
+        logger.warn("Failed to close the scanner", e);
+      }
+
+      if (failed) {
+        return;
+      }
+
+      if (fullyScanned) {
+        // If the scanner is fully scanned, we can treat it as a normal scan, and put the results
+        // into the scan set
+        snapshot.putIntoScanSet(scan, results);
+      } else {
+        // If the scanner is not fully scanned, put the results into the scanner set
+        snapshot.putIntoScannerSet(scan, results);
+      }
```

Comment:
This is a tricky part. The behavior depends on whether the storage scanner has been fully scanned.

If the storage scanner has been fully scanned, we can treat it as a normal scan and put the result into the scan set, just like a regular scan.

If the storage scanner has not been fully scanned, we put the results into the scanner set.

Reply:
> The reason why we need to differently handle the not-fully-scanned scan is that there can be remaining records?

Yes, because of that, the not-fully-scanned scan should need special handling in the serializable validation, similar to scan with limit.

> If so, how about reading the remaining records in this method and adding them to results so that we can treat the scan as a fully scanned one?

In that case, if the remaining records are large, we would consume a lot of memory. One of the advantages of the scanner API is that it only consumes memory as records are scanned. So, we should avoid doing that.

Reply:
I see. That sounds reasonable.